tests/lib/rules/jsx-one-expression-per-line.js

```diff
@@ -0,0 +1,839 @@
+/**
+ * @fileoverview Limit to one element tag per line in JSX
+ * @author Mark Ivan Allen <Vydia.com>
+ */
+
+'use strict';
+
+// ------------------------------------------------------------------------------
+// Requirements
+// ------------------------------------------------------------------------------
+
+const rule = require('../../../lib/rules/jsx-one-expression-per-line');
+const RuleTester = require('eslint').RuleTester;
+
+const parserOptions = {
+  ecmaVersion: 8,
+  sourceType: 'module',
+  ecmaFeatures: {
+    experimentalObjectRestSpread: true,
+    jsx: true
+  }
+};
+
+// ------------------------------------------------------------------------------
+// Tests
+// ------------------------------------------------------------------------------
+
+const ruleTester = new RuleTester({parserOptions});
+ruleTester.run('jsx-one-expression-per-line', rule, {
+  valid: [{
+    code: '<App />'
+  }, {
+    code: '<App></App>'
+  }, {
+    code: '<App foo="bar" />'
+  }, {
+    code: [
+      '<App>',
+      '  <Foo />',
+      '</App>'
+    ].join('\n')
+  }, {
+    code: [
+      '<App>',
+      '  <Foo />',
+      '  <Bar />',
+      '</App>'
+    ].join('\n')
+  }, {
+    code: [
+      '<App>',
+      '  <Foo></Foo>',
+      '</App>'
+    ].join('\n')
+  }, {
+    code: [
+      '<App>',
+      '  foo bar baz  whatever  ',
+      '</App>'
+    ].join('\n')
+  }, {
+    code: [
+      '<App>',
+      '  <Foo>',
+      '  </Foo>',
+      '</App>'
+    ].join('\n')
+  }, {
+    code: [
+      '<App',
+      '  foo="bar"',
+      '>',
+      '<Foo />',
+      '</App>'
+    ].join('\n')
+  }, {
+    code: [
+      '<',
+      'App',
+      '>',
+      '  <',
+      '    Foo',
+      '  />',
+      '</',
+      'App',
+      '>'
+    ].join('\n')
+  }],
+
+  invalid: [{
+    code: '<App>{"foo"}</App>',
+    output: [
+      '<App>',
+      '{"foo"}',
+      '</App>'
+    ].join('\n'),
+    errors: [{message: '`{"foo"}` must be placed on a new line'}],
+    parserOptions: parserOptions
+  }, {
+    code: '<App>foo</App>',
+    output: [
+      '<App>',
+      'foo',
+      '</App>'
+    ].join('\n'),
+    errors: [{message: '`foo` must be placed on a new line'}],
+    parserOptions: parserOptions
+  }, {
+    code: [
+      '<div>',
+      '  foo {"bar"}',
+      '</div>'
+    ].join('\n'),
+    output: [
+      '<div>',
+      '  foo ',
```

Comment:
this can cause issues due to rules that strip trailing spaces; i think this needs to remain as `foo {"bar"}`, untouched.

Reply:
`foo ` and `{"bar"}` are 2 different expressions and I don't want them on the same line.

Reply:
right - but the jsx parser doesn't ignore the space in between `foo {"bar" }` - so to make it the same, with an autofix, you'd need:
```jsx
foo
{' '}
{'bar'}
```

Reply:
Yes, if you look on the next line 117 you will see `{\' \'}`

Reply:
It results in the same output after JSX parsing. I am following the guidelines:

> Since all rules are run again after the initial round of fixes is applied, it’s not necessary for a rule to check whether the code style of a fix will cause errors to be reported by another rule.

https://eslint.org/docs/developer-guide/working-with-rules#applying-fixes

The trailing space in `foo ` is inconsequential. Its only impact is offending `no-trailing-spaces`, which, as stated in the guide, is not our concern here.